homeassistant/components/mychevy.py

```diff
@@ -0,0 +1,343 @@
+"""
+
+from logging import getLogger
+from datetime import datetime as dt
+from datetime import timedelta
+import time
+import threading
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import ENTITY_ID_FORMAT
+from homeassistant.const import (
+    CONF_USERNAME, CONF_PASSWORD)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+REQUIREMENTS = ["mychevy==0.1"]
+
+DOMAIN = 'mychevy'
+
+MYCHEVY_SUCCESS = "Success"
+MYCHEVY_ERROR = "Error"
+MYCHEVY_UNKNOWN = "Unknown"
+
+NOTIFICATION_ID = 'mychevy_website_notification'
+NOTIFICATION_TITLE = 'MyChevy website status'
+
+_LOGGER = getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=2)
+ERROR_SLEEP_TIME = 5*60
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, base_config):
+    """Setup mychevy platform."""
+    import mychevy.mychevy as mc
+
+    config = base_config.get(DOMAIN)
+    _LOGGER.debug('Received configuration: %s', config)
+
+    email = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    if hass.data.get(DOMAIN) is None:
+        hass.data[DOMAIN] = MyChevyHub(mc.MyChevy(email, password), hass)
+        hass.data[DOMAIN].start()
+
+    discovery.load_platform(hass, 'sensor', DOMAIN, {}, config)
+
+    return True
+
+
+class MyChevyHub(threading.Thread):
+    """MyChevy Hub.
+
+    Connecting to the mychevy website is done through a selenium
+    webscraping process. That can only run synchronously. In order to
+    prevent blocking of other parts of Home Assistant the architecture
+    launches a polling loop in a thread.
+
+    When new data is received, sensors are updated, and hass is
+    signaled that there are updates. Sensors are not created until the
+    first update, which will be 60 - 120 seconds after the platform
+    starts.
+    """
+
+    def __init__(self, client, hass, **kwags):
+        """Initialize MyChevy Hub."""
+        super(MyChevyHub, self).__init__()
+        self._client = client
+        self.hass = hass
+        self._car = None
+        self.add_devices = None
+        self.sensors = []
+        # This is a status sensor for the connection itself
+        self.status = MyChevyStatus(self)
+
+    @property
+    def car(self):
+        """An instance of mychevy.mychevy.EVCar."""
+        return self._car
+
+    @car.setter
+    def car(self, data):
+        """Update the EVCar.
+
+        Also update all linked sensors in hass and signal the platform
+        there are updates.
+
+        """
+        self._car = data
+        for sensor in self.sensors:
+            sensor.car = data
+            sensor.schedule_update_ha_state()
```

Comment:
Instead of calling this directly on each sensor, send a signal to each sensor via our dispatch helper. The target for the signal is the method on the entity that schedules the update.

To avoid a race where the update is called before the entity has been added to home assistant, connect the dispatcher in `async_added_to_hass` coroutine method on the entity.

Reply:
Do you have an example of that? Using this schedule call seems really straight forward, and I'm not sure I've got a clear picture of how the code would look to basically build this myself. Any examples would be appreciated.

Reply:
It's precarious to call `schedule_update_ha_state` based on an outside home assistant event, in this case when the car is updated, since you don't know if the entity has been added to home assistant yet. That's why it's good to use the dispatcher. The dispatcher allows you to decide when to connect the signal with the target. That should happen after the entity has been added to home assistant, ie in the `async_added_to_hass` method.

See eg:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/alarm_control_panel/envisalink.py#L97-L109

Note that you can implement the async version of the state update scheduling, since the dispatcher will schedule a job already on the event loop.

Since you always want to update all sensors you can have the same signal for all sensors.

Then you send the signal with `dispatcher_send`:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/dispatcher.py#L56